compiler/rustc_middle/src/ty/instance.rs

```diff
@@ -618,14 +618,15 @@ impl<'tcx> Instance<'tcx> {
         self.def.has_polymorphic_mir_body().then_some(self.substs)
     }
 
-    pub fn subst_mir<T>(&self, tcx: TyCtxt<'tcx>, v: &T) -> T
+    pub fn subst_mir<T>(&self, tcx: TyCtxt<'tcx>, v: EarlyBinder<&T>) -> T
     where
         T: TypeFoldable<TyCtxt<'tcx>> + Copy,
     {
+        let v = v.map_bound(|v| *v);
         if let Some(substs) = self.substs_for_mir_body() {
-            EarlyBinder(*v).subst(tcx, substs)
+            v.subst(tcx, substs)
         } else {
-            *v
+            v.subst_identity()
```

Comment:
Should this be `no_bound_vars().unwrap()` instead? The doc comment on `substs_for_mir_body` above seemed to suggest it might but I wasn't sure

Reply:
Yeah I think so

Reply:
So I tried `no_bound_vars().unwrap()` here and this ends up panicking when building `core` :grimacing: 

Specifically:

```plaintext
query stack during panic:
#0 [optimized_mir] optimizing MIR for `mem::maybe_uninit::<impl at library/core/src/mem/maybe_uninit.rs:270:1: 270:23>::write_slice_cloned::<impl at library/core/src/mem/maybe_uninit.rs:1128:9: 1128:42>::drop`
end of query stack
```

Re-reading the doc comment on `substs_for_mir_body`, it says (paraphrasing) "When this returns `None`, we don't want to do a substitution, since it has effectively already been done." I suppose that doesn't actually imply that there are no bound vars?

Maybe this should actually be `skip_binder`? I changed it to this in the latest commit (and in `(try_)subst_mir_and_normalize_erasing_regions`, which have a similar thing happening)